include/PyProxyHandler.hh

```diff
@@ -0,0 +1,157 @@
+   * @param vp - out-paramter for the gotten property
+   * @return true - call succeeded
+   * @return false - call failed and an exception has been raised
+   */
+  bool get(JSContext *cx, JS::HandleObject proxy,
+    JS::HandleValue receiver, JS::HandleId id,
+    JS::MutableHandleValue vp) const override;
+  /**
+   * @brief [[Set]]
+   *
+   * @param cx pointer to JSContext
+   * @param proxy The proxy object who's property we wish to set
+   * @param id Key of the property we wish to set
+   * @param v Value that we wish to set the property to
+   * @param receiver @TODO (Caleb Aikens) read ECMAScript docs about this
+   * @param result @TODO (Caleb Aikens) read ECMAScript docs about this
+   * @return true call succeed
+   * @return false call failed and an exception has been raised
+   */
+  bool set(JSContext *cx, JS::HandleObject proxy, JS::HandleId id,
+    JS::HandleValue v, JS::HandleValue receiver,
+    JS::ObjectOpResult &result) const override;
+  /**
+   * @brief [[Enumerate]]
+   *
+   * @param cx - pointer to JSContext
+   * @param proxy - The proxy object who's keys we output
+   * @param props - out-parameter of object IDs
+   * @return true - call succeeded
+   * @return false - call failed and an exception has been raised
+   */
+  bool enumerate(JSContext *cx, JS::HandleObject proxy,
+    JS::MutableHandleIdVector props) const override;
+
+  // @TODO (Caleb Aikens) The following are Spidermonkey-unique extensions, need to read into them more
+  /**
+   * @brief @TODO (Caleb Aikens) read up on what this trap does exactly
+   *
+   * @param cx pointer to JSContext
+   * @param proxy The proxy object who's property we wish to check
+   * @param id  Key of the property we wish to check
+   * @param bp out-paramter: true if object has property, false if not
+   * @return true call succeeded
+   * @return false call failed and an exception has been raised
+   */
+  bool hasOwn(JSContext *cx, JS::HandleObject proxy, JS::HandleId id,
+    bool *bp) const override;
+  /**
+   * @brief @TODO (Caleb Aikens) read up on what this trap does exactly
+   *
+   * @param cx - pointer to JSContext
+   * @param proxy - The proxy object who's keys we output
+   * @param props - out-parameter of object IDs
+   * @return true - call succeeded
+   * @return false - call failed and an exception has been raised
+   */
+  bool getOwnEnumerablePropertyKeys(
+    JSContext *cx, JS::HandleObject proxy,
+    JS::MutableHandleIdVector props) const override;
+  /**
+   * @brief Handles python object reference count when JS Proxy object is finalized
+   *
+   * @param gcx pointer to JS::GCContext
+   * @param proxy the proxy object being finalized
+   */
+  void finalize(JS::GCContext *gcx, JSObject *proxy) const override;
+
+  bool getOwnPropertyDescriptor(
+    JSContext *cx, JS::HandleObject proxy, JS::HandleId id,
+    JS::MutableHandle<mozilla::Maybe<JS::PropertyDescriptor>> desc) const override {};
```

Comment:
Instead of individually implementing the `has`,`get`,`set`,`enumerate`,`hasOwn`,`getOwnEnumerablePropertyKeys` methods, consider combining them to be handled by `getOwnPropertyDescriptor`

https://hg.mozilla.org/releases/mozilla-esr102/file/4297ce3/js/public/Proxy.h#l150
> As a convenience, a few high-level methods, like get() and set(), are given default implementations that work by calling the low-level methods, like getOwnPropertyDescriptor().

https://hg.mozilla.org/releases/mozilla-esr102/file/4297ce3/js/public/Proxy.h#l300
> The base-class implementations work by calling getOwnPropertyDescriptor() and going up the [[Prototype]] chain if necessary. The algorithm for this follows what is defined for Ordinary Objects in the ES spec. They do not follow any standard. When in doubt, override them.